src/ECharts/Monad.purs

```diff
@@ -1,70 +1,137 @@
-unDSL (DSL m) = m
+type KeyVal = Tuple String Foreign
+type Pairs = Array KeyVal
 
-derive newtype instance functorDSL ∷ Functor (DSLMonad i)
-derive newtype instance applyDSL ∷ Apply (DSLMonad i)
-derive newtype instance applicativeDSL ∷ Applicative (DSLMonad i)
-derive newtype instance bindDSL ∷ Bind (DSLMonad i)
-derive newtype instance monadDSL ∷ Monad (DSLMonad i)
+newtype CommandsT (i ∷ # Effect) m a = CommandsT (WriterT Pairs m a)
 
-type DSL i = DSLMonad i Unit
+type DSL i m = CommandsT i m Unit
+type DSL' i = CommandsT i Identity Unit
 
-set ∷ ∀ i. String → Foreign → DSL i
-set k v = DSL $ tell $ Arr.singleton $ Tuple k v
+derive instance newtypeCommandsT
+  ∷ Newtype (CommandsT i m a) _
+derive newtype instance functorDSL
+  ∷ Functor m ⇒ Functor (CommandsT i m)
+derive newtype instance applyDSL
+  ∷ Apply m ⇒ Apply (CommandsT i m)
+derive newtype instance applicativeDSL
+  ∷ Applicative m ⇒ Applicative (CommandsT i m)
+derive newtype instance bindDSL
+  ∷ Bind m ⇒ Bind (CommandsT i m)
+derive newtype instance monadDSL
+  ∷ Monad m ⇒ Monad (CommandsT i m)
+derive newtype instance monadTellCommandsT
+  ∷ Monad m ⇒ MonadTell (Array (Tuple String Foreign)) (CommandsT i m)
+derive newtype instance monadWriterCommandsT
+  ∷ Monad m ⇒ MonadWriter (Array (Tuple String Foreign)) (CommandsT i m)
+derive newtype instance plusCommandsT
+  ∷ Plus m ⇒ Plus (CommandsT i m)
+derive newtype instance altCommandsT
+  ∷ Alt m ⇒ Alt (CommandsT i m)
+derive newtype instance alternativeCommandsT
+  ∷ Alternative m ⇒ Alternative (CommandsT i m)
+derive newtype instance monadRecCommandsT
+  ∷ MonadRec m ⇒ MonadRec (CommandsT i m)
+derive newtype instance monadZeroCommandsT
+  ∷ MonadZero m ⇒ MonadZero (CommandsT i m)
+derive newtype instance monadPlusCommandsT
+  ∷ MonadPlus m ⇒ MonadPlus (CommandsT i m)
+derive newtype instance monadAffCommandsT
+  ∷ MonadAff e m ⇒ MonadAff e (CommandsT i m)
+derive newtype instance monadEffCommandsT
+  ∷ MonadEff e m ⇒ MonadEff e (CommandsT i m)
+derive newtype instance monadContCommandsT
+  ∷ MonadCont m ⇒ MonadCont (CommandsT i m)
+derive newtype instance monadThrowCommandsT
+  ∷ MonadThrow e m ⇒ MonadThrow e (CommandsT i m)
+derive newtype instance monadErrorCommandsT
+  ∷ MonadError e m ⇒ MonadError e (CommandsT i m)
+derive newtype instance monadAskCommandsT
+  ∷ MonadAsk r m ⇒ MonadAsk r (CommandsT i m)
+derive newtype instance monadReaderCommandsT
+  ∷ MonadReader r m ⇒ MonadReader r (CommandsT i m)
+derive newtype instance monadStateCommandsT
+  ∷ MonadState s m ⇒ MonadState s (CommandsT i m)
 
-get ∷ ∀ i. String → DSL i → Maybe Foreign
-get k (DSL cs) =
-  F.foldl (foldFn k) Nothing $ execWriter cs
-  where
-  foldFn ∷ String → Maybe Foreign → Tuple String Foreign → Maybe Foreign
-  foldFn k' Nothing (Tuple kk f) | k' == kk = Just f
-  foldFn _ a _ = a
+lift ∷ ∀ m i a. Monad m ⇒ m a → CommandsT i m a
+lift a = CommandsT $ WriterT do
+  x ← a
+  pure $ Tuple x []
```

Comment:
Can this not be a `lift` via `MonadTrans`?

Reply:
Idk, `MonadTrans` doesn't work neither by deriving nor by handwritten code. I think it can't handle phantom row

Reply:
You're probably right, but I'll take a quick stab at it too so at least I can understand for myself why it doesn't work.

Reply:
Seems to derive ok for me?

```
derive newtype instance monadTransCommandsT ∷ MonadTrans (CommandsT i)
```

Reply:
It derived but didn't work. That funny error from slack 
```
Error: c:/Users/zimal/Projects/purescript-echarts/src/ECharts/Commands.purs:25:6:
  No type class instance was found for

    Control.Monad.Monad t4

  The instance head contains unknown type variables. Consider adding a type annotation.

while applying a function lift
  of type MonadTrans t1 => Monad t2 => t2 t3 -> t1 t2 t3
  to argument obj
while checking that expression lift obj
  has type t0 Foreign
in value declaration series

where t0 is an unknown type
      t1 is an unknown type
      t3 is an unknown type
      t2 is an unknown type
      t4 is an unknown type
```

Reply:
It seems to be fine now: https://github.com/cryogenian/purescript-echarts/pull/1 maybe something else was changed that stopped it from breaking again.